File: frontend/src/types/schemaSnapshot.ts
// ── Schema Snapshot types ───────────────────────────────────────────────────

export interface DiffSummary {
  has_drift: boolean;
  total_differences: number;
  missing_fields: { path: string; expected_type?: string }[];
  new_fields: { path: string; actual_type?: string }[];
  type_mismatches: {
    path: string;
    expected_type?: string;
    actual_type?: string;
  }[];
}

export interface SchemaSnapshot {
  id: string;
  endpoint_id: string;
  schema_hash: string;
  field_count: number;
  diff_summary: DiffSummary | null;
  created_at: string;
}

export interface SchemaHistoryResponse {
  snapshots: SchemaSnapshot[];
}

export interface SchemaDiffResponse {
  snapshot_a_id: string;
  snapshot_b_id: string;
  diff: DiffSummary;
}

File: app/schemas/schema_snapshot.py
"""Schema Snapshot schemas — response models for the schema history API."""

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel


# ── Single snapshot response ─────────────────────────────────────────────────


class SchemaSnapshotResponse(BaseModel):
    """One schema snapshot — used in lists and detail views."""

    id: uuid.UUID
    endpoint_id: uuid.UUID
    schema_hash: str
    field_count: int
    diff_summary: dict[str, Any] | None = None
    created_at: datetime


# ── History list ─────────────────────────────────────────────────────────────


class SchemaHistoryResponse(BaseModel):
    """Chronological list of schema snapshots for an endpoint."""

    snapshots: list[SchemaSnapshotResponse]


# ── Diff comparison ──────────────────────────────────────────────────────────


class SchemaDiffResponse(BaseModel):
    """Side-by-side comparison result of two schema snapshots."""

    snapshot_a_id: uuid.UUID
    snapshot_b_id: uuid.UUID
    diff: dict[str, Any]
